pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,328 @@
+// The underlying config isn't applied until Evaluate is called.
+func NewImportConfigNode(block *ast.BlockStmt, globals ComponentGlobals, sourceType importsource.SourceType) *ImportConfigNode {
+	nodeID := BlockComponentID(block).String()
+
+	globalID := nodeID
+	if globals.ControllerID != "" {
+		globalID = path.Join(globals.ControllerID, nodeID)
+	}
+
+	cn := &ImportConfigNode{
+		nodeID:                   nodeID,
+		globalID:                 globalID,
+		label:                    block.Label,
+		componentName:            block.GetBlockName(),
+		globals:                  globals,
+		block:                    block,
+		OnBlockNodeUpdate:        globals.OnBlockNodeUpdate,
+		importChildrenUpdateChan: make(chan struct{}),
+	}
+	managedOpts := getImportManagedOptions(globals, cn)
+	cn.logger = managedOpts.Logger
+	cn.source = importsource.NewImportSource(sourceType, managedOpts, vm.New(block.Body), cn.onContentUpdate)
+	return cn
+}
+
+func getImportManagedOptions(globals ComponentGlobals, cn *ImportConfigNode) component.Options {
+	return component.Options{
+		ID:     cn.globalID,
+		Logger: log.With(globals.Logger, "config", cn.globalID),
+		Registerer: prometheus.WrapRegistererWith(prometheus.Labels{
+			"config_id": cn.globalID,
+		}, prometheus.NewRegistry()),
+		Tracer:   tracing.WrapTracer(globals.TraceProvider, cn.globalID),
+		DataPath: filepath.Join(globals.DataPath, cn.globalID),
+		GetServiceData: func(name string) (interface{}, error) {
+			return globals.GetServiceData(name)
+		},
+	}
+}
+
+// Evaluate implements BlockNode and evaluates the import source.
+func (cn *ImportConfigNode) Evaluate(scope *vm.Scope) error {
+	return cn.source.Evaluate(scope)
+}
+
+// onContentUpdate is triggered every time the managed import source has new content.
+func (cn *ImportConfigNode) onContentUpdate(importedContent string) {
+	cn.mut.Lock()
+	defer cn.mut.Unlock()
+
+	cn.inContentUpdate.Store(true)
+	defer cn.inContentUpdate.Store(false)
+
+	// If the source sent the same content, there is no need to reload.
+	if cn.importedContent == importedContent {
+		return
+	}
+
+	cn.importedContent = importedContent
+	cn.importedDeclares = make(map[string]ast.Body)
+	cn.importConfigNodesChildren = make(map[string]*ImportConfigNode)
+
+	parsedImportedContent, err := parser.ParseFile(cn.label, []byte(importedContent))
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to parse file on update", "err", err)
+		return
+	}
+
+	// populate importedDeclares and importConfigNodesChildren
+	cn.processImportedContent(parsedImportedContent)
+
+	// evaluate the importConfigNodesChildren that have been created
+	err = cn.evaluateChildren()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to evaluate nested import", "err", err)
+		return
+	}
+
+	// trigger to stop previous children from running and to start running the new ones.
+	if cn.importChildrenRunning {
+		cn.importChildrenUpdateChan <- struct{}{}
```

Comment:
Should this handle the case where the channel is already full?

```go
select {
case cn.importChildrenUpdateChan <- struct{}{}: // queued trigger
default: // trigger already queued; no-op 
}
```

(This only applies if you decide to buffer importChildrenUpdateChan, which it isn't right now)